### DISCUSSION TITLE
Is there is any code for eye blink detection using dlib c++

### DISCUSSION THREAD

**Discussion by ranjith502** [davisking/dlib#2786] [Q&A]
Help on eye blink detection using c++ dlib

**Comment by davisking** [comment]
Na, no function for that in dlib.  Just the face landmark stuff and lower level tooling.

**Comment by ranjith502** [comment]
@davisking thanks for the reply ,is there is any way or any any reference so that I can try form my side ,

**Reply by arrufat** [reply]
I guess you could refer to: http://dlib.net/dlib/image_processing/render_face_detections.h.html
This code accesses each individual landmark to draw it.
You could then check which eye key points interest you and compute some kind of distance/ratio to decide if the eye is closed.

**Reply by arrufat** [reply]
Oh, I just found this in Python, it should be straightforward to do it in C++, too.
https://pyimagesearch.com/2017/04/24/eye-blink-detection-opencv-python-dlib/

**Comment by ranjith502** [comment]
@arrufat  thanks man ,iam new to c++ dlib ,iam looking for the straight forward,any way thanks man i will learn c++ can i will modify the code thanks so much brother

**Reply by ranjith502** [reply]
@davisking @arrufat 
thanks for your help guys ,using my knowledge and with the help of chat gpt , finally a working code of detecting blink count 

this is the code >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

#include <dlib/opencv.h>
#include <opencv2/highgui/highgui.hpp>
#include <dlib/image_processing/frontal_face_detector.h>
#include <dlib/image_processing/render_face_detections.h>
#include <dlib/image_processing.h>
#include <dlib/gui_widgets.h>
#include <opencv2/imgproc/imgproc.hpp>

using namespace dlib;
using namespace std;

// Calculates the Eye Aspect Ratio (EAR) for a set of eye landmarks.
double calculate_ear(const std::vector<point>& landmarks)
{
    // Calculate the distances between the landmarks.
    double a = length(landmarks[1] - landmarks[5]);
    double b = length(landmarks[2] - landmarks[4]);
    double c = length(landmarks[0] - landmarks[3]);

    // Calculate EAR.
    double ear = (a + b) / (2.0 * c);

    return ear;
}

int main()
{
    try
    {
        cv::VideoCapture cap(0);
        if (!cap.isOpened())
        {
            cerr << "Unable to connect to camera" << endl;
            return 1;
        }

        image_window win;
        frontal_face_detector detector = get_frontal_face_detector();
        shape_predictor pose_model;
        deserialize("shape_predictor_68_face_landmarks.dat") >> pose_model;

        // Keep track of the EAR values for the previous and current frame.
        double prev_ear = 0.0;
        double curr_ear = 0.0;

        // Keep track of the number of consecutive frames where the eyes have been closed.
        int closed_frames = 0;
        int blink_count = 0;

        while (!win.is_closed())
        {
            cv::Mat temp;
            if (!cap.read(temp))
            {
                break;
            }
            cv_image<bgr_pixel> cimg(temp);

            // Detect faces in the image
            std::vector<rectangle> faces = detector(cimg);

            // cout << "Number of faces detected: " << faces.size() << endl;

            for (const auto& face : faces)
            {
                full_object_detection shape = pose_model(cimg, face);

                // Get the landmarks for the left and right eyes.
                std::vector<point> left_eye_landmarks;
                std::vector<point> right_eye_landmarks;
                for (int i = 36; i <= 41; i++)
                {
                    left_eye_landmarks.push_back(shape.part(i));
                }
                for (int i = 42; i <= 47; i++)
                {
                    right_eye_landmarks.push_back(shape.part(i));
                }

                // Calculate EAR for the left and right eyes.
                double left_ear = calculate_ear(left_eye_landmarks);
                double right_ear = calculate_ear(right_eye_landmarks);

                // Calculate the average EAR for both eyes.
                curr_ear = (left_ear + right_ear) / 2.0;

                // Print the current EAR value.
                // cout << "EAR: " << curr_ear << endl;

                // Check if the eyes are closed.
                bool closed_eyes = curr_ear < 0.2;

                if (closed_eyes)
                {
                    closed_frames++;
                    if (closed_frames >= 0)
                    {
                        blink_count++;
                        cout << "Blink detected! Total blinks: " << blink_count << endl;
                        closed_frames = 0;
                    }
                }
                else
                {
                    closed_frames = 0;
                }


                for (const auto& landmark : left_eye_landmarks)
                {
                    circle(temp, cv::Point(landmark.x(), landmark.y()), 2, cv::Scalar(0, 255, 0), -1);
                }

                // Draw circles around the landmarks of the right eye.
                for (const auto& landmark : right_eye_landmarks)
                {
                    circle(temp, cv::Point(landmark.x(), landmark.y()), 2, cv::Scalar(0, 255, 0), -1);
                }

                win.clear_overlay();
                win.set_image(cimg);
                win.add_overlay(render_face_detections(shape));
            }

        }
    }
    catch(exception& e)
    {
        cout << e.what() << endl;
    }
}
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>



it is working like a charm

**Reply by ranjith502** [reply]
@arrufat @davisking can you add this file in repo ,so that people who wants then can directly use it.

**Reply by davisking** [reply]
To be in the repo it would have to be turned into a well documented function with a clear interface.  There are also a lot of ways to do it depending on the exact need.  Probably better if people refer to the article that's already there :)

**Comment by ranjith502** [comment]
Thanks brother
